### PR TITLE
feat(port-hilserl): Proof of concept to improve episode recording UX using pygame instead of csv.imshow

### PR DESCRIPTION
**Pull request by jackvial**
## What this does
- Use pygame for rendering camera feed to screen instead of cv2.imshow. This should offer better performance, camera feed and robot movement should be smoother when recording episodes
- Use pygame event for handling keyboard inputs
- Encapsulate rendering and keyboard events in a ControlContext class that gets updated in the control_loop. This makes it simple to render helpful information on top of the camera feed e.g. current reward, control phase (warmup|record episode)

## Demo
https://github.com/user-attachments/assets/c4cb4b8e-63eb-4662-8192-6dcb90ed6174

## How To Test
`pip install pygame`

### Teleoperate
`python lerobot/scripts/control_robot.py teleoperate --robot-path lerobot/configs/robot/koch.yaml --fps 30`

Teleoperate should look like this. No episodes or rewards shown at top. 

Screenshot from 2024-12-16 21-33-26

### Record With Rewards

```
python lerobot/scripts/control_robot.py record \
    --robot-path lerobot/configs/robot/koch.yaml \
    --fps 30 \
    --root outputs/pygame_rendering_test_0 \
    --repo-id your_hf_username/pygame_rendering_test_0 \
    --warmup-time-s 2 \
    --episode-time-s 10 \
    --reset-time-s 5 \
    --num-episodes 10 \
    --push-to-hub 1 \
    --assign-rewards 1 \
    --single-task test_description
```

Warmup phase will not show rewards or episodes at the top of the screen.
Record phase will show episodes and count at top of screen.

Screenshot from 2024-12-16 21-34-46

After recording one episode I'm getting this error about the shape of the actions column, not yet sure what is happening here, will have too look into 

```
Exception has occurred: ArrowInvalid
Column 1 named action expected length 0 but got length 589
```

Screenshot from 2024-12-16 21-36-40

### CODE REVIEW

File: .cache/calibration/koch_tdmpc_jack/main_follower.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"homing_offset": [2048, 3072, 2048, -2048, -2048, 2048], "drive_mode": [1, 1, 1, 0, 0, 1], "start_pos": [2020, 3009, 1991, 2038, 2046, 2061], "end_pos": [-1034, -2102, -1055, 3011, 2986, -843], "calib_mode": ["DEGREE", "DEGREE", "DEGREE", "DEGREE", "DEGREE", "DEGREE"], "motor_names": ["shoulder_pan", "shoulder_lift", "elbow_flex", "wrist_flex", "wrist_roll", "gripper"]}
```
Comment (michel-aractingi) [inline .cache/calibration/koch_tdmpc_jack/main_follower.json:1]: Could you remove this file, calibration files are kept locally.

File: .cache/calibration/koch_tdmpc_jack/main_leader.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"homing_offset": [0, 3072, 3072, 0, 0, -2048], "drive_mode": [1, 1, 1, 0, 0, 0], "start_pos": [5, 3037, 3060, 35, -142, 1945], "end_pos": [1075, -2018, -2030, 1003, 1063, 2666], "calib_mode": ["DEGREE", "DEGREE", "DEGREE", "DEGREE", "DEGREE", "DEGREE"], "motor_names": ["shoulder_pan", "shoulder_lift", "elbow_flex", "wrist_flex", "wrist_roll", "gripper"]}
```
Comment (michel-aractingi) [inline .cache/calibration/koch_tdmpc_jack/main_leader.json:1]: Same here if you could remove please.

File: lerobot/configs/robot/koch_jack.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+_target_: lerobot.common.robot_devices.robots.manipulator.ManipulatorRobot
```
Comment (michel-aractingi) [inline lerobot/configs/robot/koch_jack.yaml:1]: Also here, this file is unnecessary.

File: record_with_rewards.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
+python lerobot/scripts/control_robot.py record \
```
Comment (michel-aractingi) [inline record_with_rewards.sh:1]: Could you remove this file but add the command in the PRs description for testing?

File: teleoperate.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#/bin/bash
+
+python lerobot/scripts/control_robot.py teleoperate --robot-path lerobot/configs/robot/koch_jack.yaml
```
Comment (michel-aractingi) [inline teleoperate.sh:3]: you can remove this file.

**Comment by jackvial** [comment]
Hey @michel-aractingi , thanks for the feedback. I've removed my scripts and configs.
I've added the commands I used for teleoperate and record along with screen shots. 

After recording one episode I get an error of
```
Exception has occurred: ArrowInvalid
Column 1 named action expected length 0 but got length 589
```

so I have either broken something or I'm missing some changes, will try figure this out tomorrow.

Also, I think `log_control_info` could also be moved to ControlContext and be enabled/disabled with a boolean flag in `ControlContextConfig`. `log_control_info` produces a lot of output and can quickly bury other information printed to the terminal so would be nice if it were configurable

**Comment by jackvial** [comment]
@michel-aractingi I've created a new branch off main for this as it will probably be easier to coordinate that way and avoid merging conflicts. This is the branch I'm working on https://github.com/jackvial/lerobot/commits/pygame-for-rendering-and-event-handling/ . I  have a bit more work to do on it but plan to open a PR in the next few days. 

Let me know if it's ok to close this draft PR

**Comment by jackvial** [comment]
closing in favor of https://github.com/huggingface/lerobot/pull/601

**Pull request closed without merging**